### PR TITLE
ICU-22403 Fix out-of-bounds in icuexportdata

### PR DESCRIPTION
**Pull request by Manishearth**
Don't read past the storageXX.size(). Refactor find-or-append into a function. Stop double-uset_close() (delete) of a set.

##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22403
- [x] Required: The PR title must be prefixed with a JIRA Issue number. 
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

**Review by sffc — APPROVED** [review]
Praise: much more clear

**Review by sffc — APPROVED** [review]
Praise: much more clear

**Comment by Manishearth** [comment]
Should we wait for @markusicu's review? Ought I squash the commits?

File: icu4c/source/tools/icuexportdata/icuexportdata.cpp
Comment (markusicu) [inline icu4c/source/tools/icuexportdata/icuexportdata.cpp:719]:
```suggestion
    // Last index where we might find the start of the complete needle.
    size_t last_start = storage.size() - needle_len;
    for (size_t i = 0; i <= last_start; i++) {
```

File: icu4c/source/tools/icuexportdata/icuexportdata.cpp
Comment (markusicu) [inline icu4c/source/tools/icuexportdata/icuexportdata.cpp:732]:
```suggestion
        for (size_t j = 0;; j++) {
            if (j == needle_len) {
                return i;  // found a match
            }
            if (storage[i + j] != uint32_t(needle[j])) {
                break;
            }
        }
```

**Review by markusicu — COMMENTED** [review]
good work!
let's simplify further.

File: icu4c/source/tools/icuexportdata/icuexportdata.cpp
Comment (markusicu) [reply icu4c/source/tools/icuexportdata/icuexportdata.cpp:719]: Actually, please make that `lastStart` not `last_start`.
And `needleLen` not `needle_len`.
ICU uses camelCase variable names.

**Comment by markusicu** [comment]
PS: The commit checker wants the commit messages to start with exactly `ICU-22403 ` -- no brackets, and exactly one ASCII space after the ticket number.

File: icu4c/source/tools/icuexportdata/icuexportdata.cpp
Comment (Manishearth) [reply icu4c/source/tools/icuexportdata/icuexportdata.cpp]: This isn't going to work because the types are unsigned

Comment (markusicu) [reply icu4c/source/tools/icuexportdata/icuexportdata.cpp]: You are right. I am not used to working with size_t. Elsewhere in ICU we always use int32_t for indexes and lengths, just like int in Java. But if you like size_t for this tools code, it's ok.

Comment (markusicu) [reply icu4c/source/tools/icuexportdata/icuexportdata.cpp]: totally optional:
```
if (needleLen <= storage.size()) {
    size_t lastStart = storage.size() - needleLen;
    for (size_t i = 0; i <= lastStart; i++) {
        ...
```

File: icu4c/source/tools/icuexportdata/icuexportdata.cpp
Comment (markusicu) [inline icu4c/source/tools/icuexportdata/icuexportdata.cpp:717]:
```suggestion
    // Last index where we might find the start of the complete needle.
```

File: icu4c/source/tools/icuexportdata/icuexportdata.cpp
Comment (markusicu) [inline icu4c/source/tools/icuexportdata/icuexportdata.cpp:730]:
```suggestion
    // We didn't find anything. Append, keeping the append index in mind.
```

**Review by markusicu — COMMENTED** [review]
lgtm pse squash

**Comment by Manishearth** [comment]
Squashed, need someone on the ICU team to merge. Once this lands I'll make a maint uplift PR

**Review by markusicu — APPROVED** [review]
lgtm tnx!

**Comment by markusicu** [comment]
rerunning failed tests

**Comment by Manishearth** [comment]
```
cp ./mappings/windows-950-2000.ucm ./out/tmp/in/mappings/windows-950-2000.ucm
LD_LIBRARY_PATH=../lib:../stubdata:../tools/ctestfw:$LD_LIBRARY_PATH  ../bin/genrb -s ./out/tmp/in/zone -d ./out/build/icudt73l/zone/ -i ./out/build/icudt73l --writePoolBundle -k de.txt de_CH.txt en.txt en_001.txt en_150.txt en_AE.txt en_AG.txt en_AI.txt en_AT.txt en_AU.txt en_BB.txt en_BE.txt en_BM.txt en_BS.txt en_BW.txt en_BZ.txt en_CA.txt en_CC.txt en_CH.txt en_CK.txt en_CM.txt en_CX.txt en_CY.txt en_DE.txt en_DG.txt en_DK.txt en_DM.txt en_ER.txt en_FI.txt en_FJ.txt en_FK.txt en_FM.txt en_GB.txt en_GD.txt en_GG.txt en_GH.txt en_GI.txt en_GM.txt en_GU.txt en_GY.txt en_HK.txt en_IE.txt en_IL.txt en_IM.txt en_IN.txt en_IO.txt en_JE.txt en_JM.txt en_KE.txt en_KI.txt en_KN.txt en_KY.txt en_LC.txt en_LR.txt en_LS.txt en_MG.txt en_MH.txt en_MO.txt en_MP.txt en_MS.txt en_MT.txt en_MU.txt en_MV.txt en_MW.txt en_MY.txt en_NA.txt en_NF.txt en_NG.txt en_NH.txt en_NL.txt en_NR.txt en_NU.txt en_NZ.txt en_PG.txt en_PK.txt en_PN.txt en_PW.txt en_RH.txt en_RW.txt en_SB.txt en_SC.txt en_SD.txt en_SE.txt en_SG.txt en_SH.txt en_SI.txt en_SL.txt en_SS.txt en_SX.txt en_SZ.txt en_TC.txt en_TK.txt en_TO.txt en_TT.txt en_TV.txt en_TZ.txt en_UG.txt en_VC.txt en_VG.txt en_VU.txt en_WS.txt en_ZA.txt en_ZM.txt en_ZW.txt hi_Latn.txt root.txt zh.txt zh_CN.txt zh_Hans.txt zh_Hans_CN.txt zh_Hans_SG.txt zh_SG.txt
LD_LIBRARY_PATH=../lib:../stubdata:../tools/ctestfw:$LD_LIBRARY_PATH  ../bin/genrb -s ./out/tmp/zone -d ./out/build/icudt73l/zone/ -i ./out/build/icudt73l -k res_index.txt
LD_LIBRARY_PATH=../lib:../stubdata:../tools/ctestfw:$LD_LIBRARY_PATH  ../bin/makeconv -s ./out/tmp/in -d ./out/build/icudt73l -c mappings/windows-950-2000.ucm
Error creating converter for "./out/build/icudt73l/windows-950-2000.cnv" file for "./out/tmp/in/mappings/windows-950-2000.ucm" (U_FILE_ACCESS_ERROR)
make[1]: *** [../data/rules.mk:13579: out/build/icudt73l/windows-950-2000.cnv] Error 4
make[1]: *** Waiting for unfinished jobs....
```

seems unrelated

**Comment by markusicu** [comment]
> seems unrelated

Yes, this one is flaky. Sometimes it fixes itself with enough reruns. Pending PR #2471 should fix this permanently.

**Comment by markusicu** [comment]
started another rerun. worst case, rebase after the other PR is in.

**Comment by markusicu** [comment]
Rerunning the DataFilter test again...

**Comment by Manishearth** [comment]
Checks pass.

**Pull request merged**
